### DISCUSSION TITLE
Version and Device Compatibility of MMKV Data

### DISCUSSION THREAD

**Discussion by JerryMu** [Tencent/MMKV#1430] [Q&A]
Is it possible for MMKV data files to be cross-platform or version-compatible? I am using POSIX-based MMKV, and the target platforms might include iOS and Android(32-bit or 64-bit) . Is it feasible to achieve resource synchronization by transferring MMKV data files?

**Comment by lingol** [comment]
MMKV is not designed to be transported between hosts, especially cross-platforms with different arches. 
For example, MMKV's file is created and extended by page size, and different OSs have different page sizes.

**Reply by JerryMu** [reply]
In this case, if I want to synchronize MMKV data across different devices, would I have to manually serialize it into a format like JSON and then convert it back to MMKV? Are there any existing better solutions?
I noticed that MMKV uses Protocol Buffers, which I believe is a reliable cross-platform format. Have you considered adding an interface to MMKV for serializing data to a file and restoring data from a file?

**Reply by lingol** [reply]
MMKV does indeed support backup and restore functionalities for the same platforms, and we encourage you to give it a try. You might find that it works seamlessly across most of your target platforms, which could be a pleasant surprise.

While we recognize the potential value in officially extending support to additional platforms, this is not currently part of our planned roadmap. However, we are always open to feedback and suggestions, and we appreciate your understanding as we continue to focus on our existing priorities.

**Reply by JerryMu** [reply]
Thank you for your suggestion. We have generated data using C++ MMKV version 1.3.9 on a Mac M1 and successfully read this data on Android devices with armv7 and armv8 architectures. Based on the current results, data generation, reading, and usage appear to be functioning correctly. However, I am aware that our testing coverage is insufficient.

Could you help us consider potential risks from a developer’s perspective when using MMKV data files across different platforms? From what we’ve seen so far, cross-platform file usage seems feasible, but it would be great to have official support and confirmation from your side in the future.

**Reply by lingol** [reply]
Well, there are several things you should test carefully. 
* The page size. As I mentioned above, different platforms have different page sizes. Normally, the page size is a multiple of 4K. While it's okay to port an MMKV from a large page-size platform to a small one, it might be problematic to port from small to large. Specifically, Win32 & Android have a 4K page size until Android 15, while iOS/macOS has 16K. I believe the `MemoryFile` class in the MMKV Core lib has managed it robustly. You should test it regardless.
* The endianness. While most of the value is encoded by variant, there's some info encoded as-is from memory. Specifically, the metafile (*.crc) is memory copied. This should be no problem if you are targeting little-endian platforms.
* The custom root path. MMKV for Android has a historical bug that mistakenly handles MMKV instances with a custom root path with an md5 internal ID and/or filename. The details are in the [code](https://github.com/Tencent/MMKV/blob/master/Core/MMKV_Android.cpp#L43). MMKV for Android has managed it correctly on backup & restoring on the same platform. You should be careful about MMKV with a custom root path.

**Comment by JerryMu** [comment]
Hello, I noticed that MMKV 2.0.2 has been released, and it mentions bug fixes related to the backup/restore functionality across different filesystems. Does this imply that the backup/restore feature itself has achieved a certain level of cross-platform capability?

Previously, I tested this on iOS x86/64 ARM as well as 32-bit and 64-bit Android projects without noticing any significant issues. However, since there was no official statement about cross-platform consistency, I ultimately chose not to use it.

I’d like to know if this update represents an effort to ensure cross-platform consistency. Looking forward to an official confirmation and test on cross-platform reliability soon.

**Reply by lingol** [reply]
The fix in v2.0.2 has nothing to do with cross-platform consistency.
